### PR TITLE
add raft node pkgs

### PR DESCRIPTION
**Pull request by upccup**

### CODE REVIEW

**Comment by pwzgorilla** [comment]
@upccup 把 如何跑raft的方法在readme里写下  怎么启动

File: raft/raft.go [hunk 1]
```diff
@@ -0,0 +1,730 @@
...
+	// indicates that the node is a raft leader
+	IsLeader LeadershipState = iota
+
+	// indicates that the node is a raft follower
+	IsFollower
+)
+
+var (
+	// returns when an operation was submiited bu the node was stopped in the meantime
+	ErrStopped = errors.New("raft: failed to process the request: node is stopped")
```
Comment (cmingxu) [inline raft/raft.go]: ErrStopped => ErrRaftNode?Server?Stopped

File: raft/raft.go [hunk 2]
```diff
@@ -0,0 +1,730 @@
...
+
+	// returns when an operation was submitted but the node lost leader status before it became committed
+	ErrLostLeadership = errors.New("raft: failed to process the request: node losts leader status")
```
Comment (cmingxu) [inline raft/raft.go]: ErrLostLeadership = ErrRaftLostLeadership

File: raft/raft.go (hunk 1 above)
Comment (cmingxu) [inline raft/raft.go]: raft: failed to process the request: node is stopped => raft: failed to process the request: node stopped

File: raft/raft.go [hunk 3]
```diff
@@ -0,0 +1,730 @@
...
+const (
+	// indicates that the node is a raft leader
+	IsLeader LeadershipState = iota
+
+	// indicates that the node is a raft follower
+	IsFollower
+)
+
+var (
+	// returns when an operation was submiited bu the node was stopped in the meantime
```
Comment (cmingxu) [inline raft/raft.go]: returns when an operation was submitted but the node stop at the same time.

File: raft/raft.go [hunk 4]
```diff
@@ -0,0 +1,730 @@
...
+	// lock should be used only in stop(), all other functions should use Rlock
+	stopMu sync.RWMutex
+
+	// waitProp waits for all proposals to be teminated before shutting down the node
+	waitProp sync.WaitGroup
+}
+
+var (
+	defaultSnapshotCount   uint64 = 10000
+	snapshotCatchUpEntrisN uint64 = 10000
```
Comment (cmingxu) [inline raft/raft.go]: snapshotCatchUpEntriesN ?

**Comment by cmingxu** [comment]
LGTM 
+1 👍

**Pull request merged**